components/PlaybackTask.brs

```diff
@@ -0,0 +1,253 @@
+Library "Roku_Ads.brs" ' Must be initialized in the task (or main) thread
+
+sub init()
+  ? "TRUE[X] >>> PlaybackTask::init()"
+  m.top.functionName = "setup"
+end sub
+
+sub setup() 
+  setupScopedVariables()
+  setupVideo()
+  setupRaf()
+  setupEvents()
+  initPlayback()
+end sub
+
+sub setupScopedVariables()
+  m.port = createObject("roMessagePort")
+  m.adFacade = m.top.adFacade
+  m.skipAds = false
+  m.lastPosition = 0
+  m.videoPlayer = invalid
+  m.raf = invalid
+  m.truexAd = invalid
+  m.currentAdPod = invalid
+end sub
+
+sub setupVideo()
+  ? "TRUE[X] >>> setupVideo()"
+
+  videoPlayer = m.top.video
+  videoContent = createObject("roSGNode", "ContentNode")
+
+  videoContent.url = "http://ctv.truex.com.s3.amazonaws.com/assets/reference-app-stream-no-ads-720p.mp4"
+  videoContent.length = 22 * 60
+
+  videoContent.title = "video title"
+  videoContent.streamFormat = "mp4"
+  videoContent.playStart = 0
+
+  videoPlayer.content = videoContent
+  videoPlayer.SetFocus(true)
+  videoPlayer.visible = true
+  videoPlayer.observeField("position", m.port)
+  videoPlayer.EnableCookies()
+
+  m.videoPlayer = videoPlayer
+end sub
+
+sub setupRaf()
+  ? "TRUE[X] >>> setupRaf()"
+  raf = Roku_Ads()
+  raf.enableAdMeasurements(true)
+  raf.setContentGenre("Entertainment")
+  raf.setContentId("TrueXSample")
+
+  raf.SetDebugOutput(false) 'debugging
+  raf.SetAdPrefs(false)
+
+  adUrl = m.top.adUrl
+  if adUrl = invalid OR adUrl = ""
+    ' adUrl = "pkg:/res/adpods/vast-regular.xml"  ' Regular Ads
+    adUrl = "pkg:/res/adpods/vmap-truex.xml" ' Preroll + Midroll Truex experience
+    ' adUrl = "pkg:/res/adpod/truex-pod-preroll.xml"
+    ' adUrl = "pkg:/res/adpod/truex-pod-skip.xml" ' Forced standalone skip pod.  Can be inserted in a full ad payload
+  end if 
+  raf.setAdUrl(adUrl)
+
+  m.raf = raf
+end sub
+
+sub setupEvents()
+  m.top.observeField("exitPlayback", m.port)
+end sub
+
+sub initPlayback()
+  ? "TRUE[X] >>> initPlayback()"
+
+  m.currentAdPod = getPreroll()
+  startPlayback() ' Will handle playing a preroll if it exists per above
+
+  while(true)
+    msg = Wait(0, m.port)
+    msgType = type(msg)
+
+    ' ? "event type:" msgType
+    if msgType = "roSGNodeEvent"
+      field = msg.getField()
+      ' ? "roSGNodeEvent msg.getField()" field
+      if field = "position" then 
+        onPositionChanged(msg)
+      else if field = "event" then
+        onTrueXEvent(msg)
+      else if field = "exitPlayback"
+        exitPlayback()
+      end if
+      ' TODO: Video complete event
```

Comment:
Scrubbed the comment to avoid confusion.  If this is desired behaviour for the ref app, let me know.  I figure it is listening for an event, or checking if the position is at the end, then exit the player.